Code/Framework/AzCore/AzCore/DOM/DomValue.h

```diff
@@ -0,0 +1,385 @@
+        float GetFloat() const;
+        void SetFloat(float);
+
+        // String API...
+        AZStd::string_view GetString() const;
+        size_t GetStringLength() const;
+        void SetString(AZStd::string_view);
+        void SetString(AZStd::shared_ptr<const AZStd::string>);
+        void CopyFromString(AZStd::string_view);
+
+        // Opaque type API...
+        AZStd::any& GetOpaqueValue() const;
+        //! This sets this Value to represent a value of an type that the DOM has
+        //! no formal knowledge of. Where possible, it should be preferred to
+        //! serialize an opaque type into a DOM value instead, as serializers
+        //! and other systems will have no means of dealing with fully arbitrary
+        //! values.
+        void SetOpaqueValue(AZStd::any&);
+
+        // Null API...
+        void SetNull();
+
+        // Visitor API...
+        Visitor::Result Accept(Visitor& visitor, bool copyStrings) const;
+        AZStd::unique_ptr<Visitor> GetWriteHandler();
+
+        bool DeepCompareIsEqual(const Value& other) const;
+        Value DeepCopy(bool copyStrings = true) const;
+
+    private:
+        const Node& GetNodeInternal() const;
+        Node& GetNodeInternal();
+        const Object::ContainerType& GetObjectInternal() const;
+        Object::ContainerType& GetObjectInternal();
+        const Array::ContainerType& GetArrayInternal() const;
+        Array::ContainerType& GetArrayInternal();
+
+        explicit Value(AZStd::any* opaqueValue);
+
+        // Determine the short string buffer size based on the size of our largest internal type (string_view)
+        // minus the size of the short string size field.
+        static constexpr const size_t ShortStringSize = sizeof(AZStd::string_view) - sizeof(size_t);
+        struct ShortStringType
+        {
+            AZStd::array<char, ShortStringSize> m_data;
+            size_t m_size;
+
+            bool operator==(const ShortStringType& other) const
+            {
+                return m_size == other.m_size ? memcmp(m_data.data(), other.m_data.data(), m_size) == 0 : false;
+            }
+        };
+
+        //! The internal storage type for Value.
+        //! These types do not correspond one-to-one with the Value's external Type as there may be multiple storage classes
+        //! for the same type in some instances, such as string storage.
+        using ValueType = AZStd::variant<
+            // NullType
+            AZStd::monostate,
+            // NumberType
+            int64_t,
+            uint64_t,
+            double,
+            // FalseType & TrueType
+            bool,
+            // StringType
+            AZStd::string_view,
+            AZStd::shared_ptr<const AZStd::string>,
+            ShortStringType,
+            // ObjectType
+            ObjectPtr,
+            // ArrayType
+            ArrayPtr,
+            // NodeType
+            NodePtr,
+            // OpaqueType
+            AZStd::any*>;
```

Comment:
Why is the string wrapped in a `shared_ptr` but this `AZStd::any` is a raw pointer? Would it be useful to wrap this in a `shared_ptr` since it's arbitrary data so it can be used in unpredictable ways, or at least wrap it in a `AZStd::unique_ptr` to make sure there's only ever one owner?

Reply:
Opaque storage is now `AZStd::shared_ptr<AZStd::any>` to mitigate lifecycle concerns.